thrust/system/cuda/detail/adjacent_difference.h

```diff
@@ -443,27 +188,49 @@ namespace __adjacent_difference {
                       OutputIt                   result,
                       BinaryOp                   binary_op)
   {
-    typedef typename iterator_traits<InputIt>::difference_type size_type;
-
-    size_type    num_items    = thrust::distance(first, last);
-    size_t       storage_size = 0;
-    cudaStream_t stream       = cuda_cub::stream(policy);
-    bool         debug_sync   = THRUST_DEBUG_SYNC_FLAG;
-
-    cudaError_t status;
-    THRUST_INDEX_TYPE_DISPATCH(status, doit_step, num_items,
-        (NULL, storage_size, first, result, binary_op,
-           num_items_fixed, stream, debug_sync));
+    const auto num_items =
+      static_cast<std::size_t>(thrust::distance(first, last));
+    std::size_t storage_size = 0;
+    cudaStream_t stream = cuda_cub::stream(policy);
+    const bool debug_sync = THRUST_DEBUG_SYNC_FLAG;
+
+    using UnwrapInputIt = thrust::detail::try_unwrap_contiguous_iterator_return_t<InputIt>;
+    using UnwrapOutputIt = thrust::detail::try_unwrap_contiguous_iterator_return_t<OutputIt>;
+
+    constexpr bool can_compare_iterators =
+      is_contiguous_iterator<UnwrapInputIt>::value &&
```

Comment:
Suggestion: This should be `is_pointer` -- it the `InputIt` was contiguous, the `UnwrapInputIt` is guaranteed to be a pointer here.

I don't think it'd break anything here to use `in_contiguous`, but I'd find the intent clearer if we explicitly checked for pointer-ness -- especially since the `in_place` checks are comparing the iterators (e.g. `first_unwrap != result_unwrap`) and not the memory addresses of the iterators' data (e.g. `&*first_unwrap != &* result_unwrap`).